docs/design/CONFIG_API.md

```diff
@@ -0,0 +1,330 @@
+* Define a complete configuration API for managing Envoy Gateway at runtime. The configuration API will evolve over
+  time.
+* Define a complete configuration API for managing control and data planes. The configuration API will evolve over time.
+
+## Proposal
+
+Utilize the [Kubebuilder][kubebuilder] framework to build and manage Envoy Gateway APIs. Kubebuilder facilitates the
+following developer workflow for building APIs:
+
+- Create one or more resource APIs as CRDs and then add fields to the resources.
+- Implement reconcile loops in controllers and watch additional resources.
+- Test by running against a cluster (self-installs CRDs and starts controllers automatically).
+- Update bootstrapped integration tests to test new fields and business logic.
+- Build and publish a container from a Dockerfile.
+
+After installing the latest Kubebuilder release, create the `RuntimeConfig` API:
+```shell
+kubebuilder create api --group gateway.envoyproxy.io --version v1alpha1 --kind RuntimeConfig
+```
+
+The `v1alpha1` version and `gateway.envoyproxy.io` API group get generated:
+```go
+// gateway/api/v1alpha1/doc.go
+
+// Package v1alpha1 contains API Schema definitions for the gateway.envoyproxy.io API group.
+//
+// +groupName=gateway.envoyproxy.io
+package v1alpha1
+```
+
+The initial `RuntimeConfig` API being proposed:
+```go
+// gateway/api/v1alpha1/runtimeconfig_types.go
+
+// RuntimeConfigSpec defines the desired state of RuntimeConfig.
+type RuntimeConfigSpec struct {
+	// Providers provide runtime configuration. If unspecified, the Kubernetes
+	// provider will be used with default parameters.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=4
+	// +kubebuilder:default={kubernetes: {foo: bar}}
+	Providers []ProviderConfig `json:"providers,omitempty"`
+
+	// ControlPlane defines control plane configuration parameters. If unspecified,
+	// the control plane will use default configuration parameters.
+	//
+	// +optional
+	ControlPlane *ControlPlaneConfig `json:"controlPlane,omitempty"`
+
+	// DataPlane defines data plane configuration parameters. If unspecified,
+	// the data plane will use default configuration parameters.
+	//
+	// +optional
+	DataPlane *DataPlaneConfig `json:"dataPlane,omitempty"`
+}
+
+// ProviderConfig defines runtime configuration providers.
+type ProviderConfig struct {
+	// Kubernetes defines the configuration of the Kubernetes provider. Kubernetes
+	// provides runtime configuration via the Kubernetes API.
+	//
+	// +optional
+	Kubernetes KubernetesConfig `json:"kubernetes,omitempty"`
+
+	// File defines the configuration of the File provider. File provides runtime
+	// configuration defined by one or more files.
+	//
+	// +optional
+	File FileConfig `json:"file,omitempty"`
+
+	// HTTP defines the configuration of the HTTP(S) provider. HTTP provides runtime
+	// configuration by defining an HTTP(S) endpoint.
+	//
+	// +optional
+	HTTP HTTPConfig `json:"http,omitempty"`
```

Comment:
can we rm HTTP for now, since it will need an explicit design and might delay this design